src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs

```diff
@@ -11,60 +11,72 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+            public RuleSetFile(string filePath, VisualStudioRuleSetManager ruleSetManager)
             {
-                _filePath = filePath;
+                FilePath = filePath;
                 _ruleSetManager = ruleSetManager;
+            }
 
-                ImmutableArray<string> includes;
-
-                try
-                {
-                    includes = RuleSet.GetEffectiveIncludesFromFile(filePath);
-                }
-                catch (Exception e)
+            public void InitializeFileTracking(IVsFileChangeEx fileChangeService)
+            {
+                lock (_gate)
                 {
-                    // We couldn't read the rule set for whatever reason. Capture the exception
-                    // so we can surface the error later, and subscribe to file change notifications
-                    // so that we'll automatically reload the file if the user can fix the issue.
-                    _optionsRead = true;
-                    _specificDiagnosticOptions = ImmutableDictionary<string, ReportDiagnostic>.Empty;
-                    _exception = e;
-
-                    includes = ImmutableArray.Create(filePath);
-                }
+                    if (_trackers == null)
+                    {
+                        ImmutableArray<string> includes;
 
-                _trackers = new List<FileChangeTracker>(capacity: includes.Length);
+                        try
+                        {
+                            includes = RuleSet.GetEffectiveIncludesFromFile(FilePath);
+                        }
+                        catch (Exception e)
+                        {
+                            // We couldn't read the rule set for whatever reason. Capture the exception
+                            // so we can surface the error later, and subscribe to file change notifications
+                            // so that we'll automatically reload the file if the user can fix the issue.
+                            _optionsRead = true;
+                            _specificDiagnosticOptions = ImmutableDictionary<string, ReportDiagnostic>.Empty;
+                            _exception = e;
+
+                            includes = ImmutableArray.Create(FilePath);
+                        }
 
-                foreach (var include in includes)
-                {
-                    var tracker = new FileChangeTracker(fileChangeService, include);
-                    tracker.UpdatedOnDisk += IncludeUpdated;
-                    tracker.StartFileChangeListeningAsync();
+                        _trackers = new List<FileChangeTracker>(capacity: includes.Length);
+
+                        foreach (var include in includes)
+                        {
+                            var tracker = new FileChangeTracker(fileChangeService, include);
+                            tracker.UpdatedOnDisk += IncludeUpdated;
+                            tracker.StartFileChangeListeningAsync();
```

Comment:
💭 I'm concerned about this being called inside the lock, since it's generally not good to assume the call will execute asynchronously.

Reply:
The API is actually void returning -- we call it Async to indicate it _doesn't_ do synchronously. I expect some larger file change watching rewrite to happen in a month or two (or part of the bigger change) so this might get a lot better.

Incidently this wasn't a new problem for this code: if you see the handler for the file change, you'll see it was doing a workarould already. The lock was just invisible: the UI thead was the implicit lock.